core/client/src/main/java/alluxio/client/block/RetryHandlingBlockWorkerClient.java

```diff
@@ -61,41 +65,50 @@
   private final InetSocketAddress mWorkerDataServerAddress;
   private final WorkerNetAddress mWorkerNetAddress;
   private final InetSocketAddress mRpcAddress;
-  private final ExecutorService mExecutorService;
-  private final HeartbeatExecutor mHeartbeatExecutor;
-  private final Future<?> mHeartbeat;
+  private static final ScheduledExecutorService HEARTBEAT_POOL = new ScheduledThreadPoolExecutor(
+      Configuration.getInt(PropertyKey.USER_BLOCK_WORKER_CLIENT_THREADS),
+      ThreadFactoryUtils.build("block-worker-heartbeat-%d", true));
+  private static final ExecutorService HEARTBEAT_CANCEL_POOL = Executors.newFixedThreadPool(5,
+      ThreadFactoryUtils.build("block-worker-heartbeat-cancel-%d", true));
+
+  private ScheduledFuture<?> mHeartbeat = null;
 
   /**
    * Creates a {@link RetryHandlingBlockWorkerClient}. Set sessionId to null if no session Id is
    * required when using this client. For example, if you only call RPCs like promote, a session
    * Id is not required.
    *
    * @param workerNetAddress to worker's location
-   * @param executorService the executor service
    * @param sessionId the id of the session
    * @throws IOException if it fails to register the session with the worker specified
    */
-  public RetryHandlingBlockWorkerClient(WorkerNetAddress workerNetAddress,
-      ExecutorService executorService, Long sessionId) throws IOException {
+  public RetryHandlingBlockWorkerClient(WorkerNetAddress workerNetAddress, Long sessionId)
+      throws IOException {
     mRpcAddress = NetworkAddressUtils.getRpcPortSocketAddress(workerNetAddress);
 
     mWorkerNetAddress = Preconditions.checkNotNull(workerNetAddress);
     mWorkerDataServerAddress = NetworkAddressUtils.getDataPortSocketAddress(workerNetAddress);
-    mExecutorService = executorService;
     mSessionId = sessionId;
-    if (sessionId != null && executorService != null) {
-      mHeartbeatExecutor = new BlockWorkerClientHeartbeatExecutor(this);
+    if (sessionId != null) {
+      mHeartbeat = HEARTBEAT_POOL.scheduleAtFixedRate(new Runnable() {
```

Comment:
Could you add some more comment here regarding what we discussed offline: e.g., heartbeats are scheduled at per-heartbeat op basis at the fixed given rate, and heartbeating to different block workers can share the same thread.